lib/OpenQA/Worker/Jobs.pm

```diff
@@ -91,27 +91,37 @@ sub _kill_worker($) {
 sub start_job;
 
 sub check_job {
-    state $running;
-    remove_timer('check_job');
-    return if $running;
-    return unless verify_workerid;
-    $running = 1;
-    if (!$job) {
-        print "checking for job ...\n" if $verbose;
-        my $res = api_call('post', "workers/$workerid/grab_job", $worker_caps) || {job => undef};
-        $job = $res->{job};
-        if ($job && $job->{id}) {
-            Mojo::IOLoop->next_tick(\&start_job);
-        }
-        else {
-            $job = undef;
-        }
-    }
-    $running = 0;
+    my (@todo) = @_;
+    state $check_job_running;
+    return unless @todo;
+    my $host = shift @todo;
+    return if $check_job_running->{$host};
+    return unless my $workerid = $hosts->{$host}{workerid};
+    return if $job;
+    $check_job_running->{$host} = 1;
+    print "checking for job with webui $host ...\n" if $verbose;
+    api_call(
+        'post',
+        "workers/$workerid/grab_job",
+        params   => $worker_caps,
+        host     => $host,
+        callback => sub {
+            my ($res) = @_;
+            return unless ($res && $res->{job});
+            $job = $res->{job};
+            if ($job && $job->{id}) {
+                Mojo::IOLoop->next_tick(sub { start_job($host) });
+            }
+            else {
+                $job = undef;
+                Mojo::IOLoop->next_tick(sub { check_job(@todo) });
+            }
+            $check_job_running->{$host} = 0;
+        });
```

Comment:
At the very least an empty line after `my (@todo) = @_;`, and one  before the `api_call`. 😉